File: GeneradorTxt.py
# Generador de TXT
# 3 (cantidad de activos)
# Activo1 1000.0 0.05 0.2
# Activo2 1500.0 0.03 0.15
# Activo3 1200.0 0.07 0.25

import random

def GeneradorTxt(cantidad):
    try:
        with open("datos.txt", "w") as f:
            f.write(str(cantidad) + "\n")
            for i in range(cantidad):
                amount = random.uniform(1000, 2000)
                return_rate = random.uniform(0.01, 0.1)
                risk_level = random.uniform(0.1, 0.3)
                f.write(f"Activo{i+1} {amount:.2f} {return_rate:.2f} {risk_level:.2f}\n")
        print("Archivo generado con exito")
    except Exception as e:
        print(f"Error al generar el archivo: {e}")

GeneradorTxt(500)

File: GeneradorTxt.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <omp.h>

void GeneradorTxt(int cantidad) {
    // Reservar memoria para almacenar los datos de los activos
    char **datos = malloc(cantidad * sizeof(char *));
    if (datos == NULL) {
        printf("Error al reservar memoria.\n");
        return;
    }
    for (int i = 0; i < cantidad; i++) {
        datos[i] = malloc(50 * sizeof(char));
        if (datos[i] == NULL) {
            printf("Error al reservar memoria para el activo %d.\n", i + 1);
            return;
        }
    }

    srand(time(0)); // Inicializar la semilla para los números aleatorios

    // Generar los datos en paralelo
    #pragma omp parallel
    {
        int id = omp_get_thread_num();
        unsigned int seed = time(0) + id * 37; // Semilla única para cada hilo

        #pragma omp for
        for (int i = 0; i < cantidad; i++) {
            seed = seed * 1103515245 + 12345; // Algoritmo para generar una nueva semilla
            double amount = (seed % 1001) + 1000; // Generar un número entre 1000 y 2000
            seed = seed * 1103515245 + 12345; // Nueva semilla
            double return_rate = ((seed % 100) / 1000.0) + 0.01; // Generar un número entre 0.01 y 0.1
            seed = seed * 1103515245 + 12345; // Nueva semilla
            double risk_level = ((seed % 200) / 1000.0) + 0.1; // Generar un número entre 0.1 y 0.3

            // Almacenar en el buffer local de datos
            sprintf(datos[i], "Activo%d %.2f %.2f %.2f\n", i + 1, amount, return_rate, risk_level);
        }
    }

    // Escribir todos los datos generados en el archivo
    FILE *file = fopen("datos.txt", "w");
    if (file == NULL) {
        printf("Error al abrir el archivo.\n");
        return;
    }

    fprintf(file, "%d\n", cantidad);

    for (int i = 0; i < cantidad; i++) {
        fprintf(file, "%s", datos[i]);
    }

    fclose(file);
    printf("Archivo generado con éxito.\n");

    // Liberar memoria
    for (int i = 0; i < cantidad; i++) {
        free(datos[i]);
    }
    free(datos);
}

int main() {
    int cantidad = 500;
    GeneradorTxt(cantidad);
    return 0;
}
